### DISCUSSION TITLE
during initrdflash: new 'extra-pre-wipe' script

### DISCUSSION THREAD

**Discussion by Red-Dragon-99** [OE4T/meta-tegra#1648] [Ideas]
Hi,
I'd like to propose a new "**extra-pre-wipe**" feature to run scripts during initrd-flashing, similar to the already existing "extra" scripts feature:

I want to keep some files persistent to make them survive even initrd flashing. For that I want to run a safe-script before wiping the SSD and then a restore-script just after flashing the data to the SSD has finished to bring these files persistent files back. 
For the second script, I use the 'extra' scripts feature in the init-flash.sh. And to run the first script (the safe-script), I made a small modification to init-flash.sh:

```
diff --git a/init-flash.sh b/init-flash.sh
index c1cd4cfc..f7489273 100644
--- a/init-flash.sh
+++ b/init-flash.sh
@@ -165,6 +165,13 @@ else
 		    echo "/dev/nvme0n1 does not exist, skipping" > /tmp/flashpkg/flashpkg/logs/erase-nvme.log
 		fi
 		;;
+	    extra-pre-wipe)
+		if [ -f "/init-extra-pre-wipe" ]; then
+		    ./init-extra-pre-wipe
+		else
+		    echo "No init-extra-pre-wipe was found" >&2
+		fi
+		;;
 	    export-devices)
 		for dev in $args; do
 		    if setup_usb_export /dev/$dev $dev 2>&1 > /tmp/flashpkg/flashpkg/logs/export-$dev.log; then
```

----------
With init-extra-pre-wipe similar to init-extra:

```
#!/bin/sh

# Run custom extra pre wipe init script(s) provided from seperate recipes.

scripts_dir="/init-extra-pre-wipe.d"

if [ -d "${scripts_dir}" ] && [ "$(ls -A ${scripts_dir})" ]; then
  for script in "${scripts_dir}"/*; do
    ./"${script}" 2>&1 | tee "/tmp/flashpkg/flashpkg/logs/custom-extra-pre-wipe-script-$(basename "${script}").log"
  done
else
  echo "No script found in ${scripts_dir}, ignoring" > /tmp/flashpkg/flashpkg/logs/custom-script.log
fi
```

Adding the "extra-pre-wipe" is quite simple, invoke it before filling the disk with "export-devices":

```
diff --git a/initrd-flash.sh b/initrd-flash.sh
index 67756157..5f9bf0c1 100644
--- a/initrd-flash.sh
+++ b/initrd-flash.sh
@@ -364,6 +364,7 @@ generate_flash_package() {
     if [ $erase_nvme -eq 1 ]; then
 	echo "erase-nvme" >> "$mnt/flashpkg/conf/command_sequence"
     fi
+    echo "extra-pre-wipe" >> "$mnt/flashpkg/conf/command_sequence"
     if [ $EXTERNAL_ROOTFS_DRIVE -eq 1 -a $BOOT_PARTITIONS_ON_EMMC -eq 1 ]; then
 	echo "export-devices mmcblk0 $ROOTFS_DEVICE" >> "$mnt/flashpkg/conf/command_sequence"
     else
```

I just realize, one could also put that one line change before the if [ $erase-nvme -eq 1 ] block, but that one is never triggered when starting a "normal" initrd-flash via _sudo ./initrd-flash_

Any comments are welcome.

Regards
Stefan

**Comment by madisongh** [comment]
Having a hook for customizing the process at an early stage sounds good to me.

>  one could also put that one line change before the if [ $erase-nvme -eq 1 ] block

Yep, I'd move it above, so it could be used to stash data that might be on the NVMe drive before anything else happens, for folks that need to use the `--erase-nvme` option on `initrd-flash`.
